From 86b80b4442027c2fa4f22bb3aa602abb0e15ff85 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Mon, 7 Oct 2024 14:37:13 +0200
Subject: [PATCH] bootlocptr: reduce size of pbi section

- Remove duplicate 'loadc' commands, the condition register value is not
  affected by 'jump' and 'jumpc'. Saves 6 words.
- Remove 'write' command for bootlocptr high byte, its value zero is
  shared by all boot sources and implicitly initialised. Saves 6 words.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 lx2160asi/bootlocptr_auto.rcw | 31 +++++++++++--------------------
 1 file changed, 11 insertions(+), 20 deletions(-)

diff --git a/lx2160asi/bootlocptr_auto.rcw b/lx2160asi/bootlocptr_auto.rcw
index 24c7286..08e7916 100644
--- a/lx2160asi/bootlocptr_auto.rcw
+++ b/lx2160asi/bootlocptr_auto.rcw
@@ -5,56 +5,47 @@
  *
  * For single boot-source atf create_pbl adds block copy commands
  * and sets boot location pointer.
- * Automatic boot relies on rcw to include those commands;
- * create_pbl merely replaces their arguments.
+ * Automatic boot relies on rcw to include the blockcopy and bootlocptr
+ * commands, create_pbl then replaces their arguments.
  *
  * Copyright 2020 Rabeeh Khoury <rabeeh@solid-run.com>
  * Copyright 2024 Josua Mayer <josua@solid-run.com>
  *
  * Changelog:
  * - 28/09/2024: changed formatting and comments
+ * - 07/10/2024: removed unnecessary commands to reduce size
  */
 .pbi
 /* Load condition PORSR1 and mask RCW_SRC */
 loadc 0x01e00000,0x07800000
 
-/* If it is 0x8 << 23 (SDHC1) then skip the following jump command */
+/* If condition is 0x8 << 23 (SDHC1) then skip the following jump command */
 jumpc 0x00000014,0x04000000
 
 /* skip sdhc1 boot */
-jump 0x28 /* this jump + blockcopy + write = (4+4+4+4)+(4+4)+2x(4+4)=40 bytes */
+jump 0x20 /* this jump + blockcopy = (4+4+4+4)+(4+4)=24 bytes */
+
 /* copy blocks from sdhc1 (atf create_pbl will fixup arguments) */
 blockcopy 0x08,0x0000a000,0x1800d000,0x00020000
-
-/* set boot location pointer for sdhc */
 write 0x01e00400,0x1800d000
-write 0x01e00404,0x00000000
 
-/* If it is 0x9 << 23 (SDHC2) then skip the following jump command */
-loadc 0x01e00000,0x07800000
+/* If condition is 0x9 << 23 (SDHC2) then skip the following jump command */
 jumpc 0x00000014,0x04800000
 
 /* skip sdhc2 boot */
-jump 0x28 /* this jump + blockcopy + write = (4+4+4+4)+(4+4)+2x(4+4)=40 bytes */
+jump 0x20 /* this jump + blockcopy = (4+4+4+4)+(4+4)=24 bytes */
 
 /* copy blocks from sdhc1 (atf create_pbl will fixup arguments) */
 blockcopy 0x09,0x0000a000,0x1800d000,0x00020000
-
-/* set boot location pointer for sdhc */
 write 0x01e00400,0x1800d000
-write 0x01e00404,0x00000000
 
-/* If it is 0xf << 23 (XSPI) then skip the following jump command */
-loadc 0x01e00000,0x07800000
+/* If condition is 0xf << 23 (XSPI) then skip the following jump command */
 jumpc 0x00000014,0x07800000
 
 /* skip xspi boot */
-jump 0x28 /* this jump + blockcopy + write = (4+4+4+4)+(4+4)+2x(4+4)=40 bytes */
+jump 0x20 /* this jump + blockcopy = (4+4+4+4)+(4+4)=24 bytes */
 
 /* copy blocks from xspi (atf create_pbl will fixup arguments) */
 blockcopy 0x0f,0x20009000,0x1800d000,0x00020000
-
-/* set boot location pointer for xspi */
-write 0x01e00400,0x20100000
-write 0x01e00404,0x00000000
+write 0x01e00400,0x1800d000
 .end
-- 
2.43.0

